batcher/aligned-batcher/src/retry.rs

```diff
@@ -0,0 +1,317 @@
+use backon::ExponentialBuilder;
+use backon::Retryable;
+use ethers::prelude::*;
+use log::warn;
+use std::{future::Future, time::Duration};
+
+use crate::eth::payment_service::BatcherPaymentService;
+
+pub const DEFAULT_MIN_DELAY: u64 = 2000;
+pub const DEFAULT_MAX_TIMES: usize = 3;
+pub const DEFAULT_FACTOR: f32 = 2.0;
+
+#[derive(Debug)]
+pub enum RetryError<E> {
+    Transient(E),
+    Permanent(E),
+}
+
+impl<E: std::fmt::Display> std::fmt::Display for RetryError<E> {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        match self {
+            RetryError::Transient(e) => write!(f, "{}", e),
+            RetryError::Permanent(e) => write!(f, "{}", e),
+        }
+    }
+}
+
+impl<E> RetryError<E> {
+    pub fn inner(self) -> E {
+        match self {
+            RetryError::Transient(e) => e,
+            RetryError::Permanent(e) => e,
+        }
+    }
+}
+
+impl<E: std::fmt::Display> std::error::Error for RetryError<E> where E: std::fmt::Debug {}
+
+pub async fn retry_function<FutureFn, Fut, T, E>(
+    function: FutureFn,
+    min_delay: u64,
+    factor: f32,
+    max_times: usize,
+) -> Result<T, RetryError<E>>
+where
+    Fut: Future<Output = Result<T, RetryError<E>>>,
+    FutureFn: FnMut() -> Fut,
+{
+    let backoff = ExponentialBuilder::default()
+        .with_min_delay(Duration::from_millis(min_delay))
+        .with_max_times(max_times)
+        .with_factor(factor);
+
+    function
+        .retry(backoff)
+        .sleep(tokio::time::sleep)
+        .when(|e| matches!(e, RetryError::Transient(_)))
+        .await
+}
+
+pub async fn get_user_balance_retryable(
+    payment_service: &BatcherPaymentService,
+    payment_service_fallback: &BatcherPaymentService,
+    addr: &Address,
+) -> Result<U256, RetryError<String>> {
+    if let Ok(balance) = payment_service.user_balances(*addr).call().await {
+        return Ok(balance);
+    };
+
+    payment_service_fallback
+        .user_balances(*addr)
+        .call()
+        .await
+        .map_err(|e| {
+            warn!("Failed to get balance for address {:?}. Error: {e}", addr);
+            RetryError::Transient(e.to_string())
+        })
+}
+
+pub async fn get_user_nonce_from_ethereum_retryable(
```

Comment:
Adding these functions logics here seems odd.
Maybe the outer function could be in this file, which contains the call to the core function with the retries; and the core function that contains the logic could go back to `lib.rs`

Reply:
I agree, this doesn't seem to be the right place for the business logic. The mechanism belongs here.

Reply:
I'm inclined to guess they're here for the tests. You can just import the functions inside `mod test`, while placing them near their call sites.

Reply:
Moved the retryable functions to `retry::batcher_retry` module